src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,420 @@
+        private final byte[] info;
+        private final int length;
+
+        /**
+         * Constructor that may be used to initialize an {@code Expand} object
+         *
+         * @param prk
+         *     the pseudorandom key; may be {@code null}
+         * @param info
+         *     the optional context and application specific information
+         * @param length
+         *     the length of output keying material
+         */
+        private Expand(SecretKey prk, byte[] info, int length) {
+            // a null prk could be indicative of ExtractExpand
+            this.prk = prk;
+            this.info = (info == null) ? null : info.clone();
+            if (length < 1) {
+                throw new IllegalArgumentException("length must be >= 1");
+            }
+            this.length = length;
+        }
+
+        /**
+         * Returns the pseudorandom key.
+         *
+         * @return the pseudorandom key
+         */
+        public SecretKey prk() {
+            return prk;
+        }
+
+        /**
+         * Returns the optional context and application specific information.
+         *
+         * @return the optional context and application specific information
+         */
+        public byte[] info() {
+            return (info == null)? null : info.clone();
+        }
+
+        /**
+         * Returns the length of output keying material.
+         *
+         * @return the length of output keying material
+         */
+        public int length() {
+            return length;
+        }
+
+    }
+
+    /**
+     * Defines the input parameters of an ExtractExpand operation as defined in
+     * <a href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    final class ExtractExpand implements HKDFParameterSpec {
+        private final Extract ext;
+        private final Expand exp;
+
+        /**
+         * Constructor that may be used to initialize an {@code ExtractExpand}
+         * object
+         * <p>
+         * Note: {@code addIKMValue} and {@code addSaltValue} may be called
+         * afterward to supply additional values, if desired
+         *
+         * @param ext
+         *     a pre-generated {@code Extract}
+         * @param info
+         *     the optional context and application specific information (may be {@code null})
+         * @param length
+         *     the length of output keying material
+         */
+        private ExtractExpand(Extract ext, byte[] info, int length) {
+            if (ext == null) {
+                throw new IllegalArgumentException(
+                    "ext (the Extract parameter) must not be null");
+            } else {
+                this.ext = ext;
+            }
+            if (length < 1) {
+                throw new IllegalArgumentException("length must be >= 1");
+            }
+            this.exp = expand(null, info, length);
+        }
+
+        /**
+         * Returns the {@code List} of initial key material values.
```

Comment:
Can the ikms and salts method specify the ordering of the list? I assume it's the order that the material and sats were added but the API docs doesn't say this.

Reply:
I agree - it would be good to state the order.